src/Microsoft.Experimental.Collections/Microsoft/Collections/Extensions/DictionarySlim.cs

```diff
@@ -0,0 +1,284 @@
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private int GetBucketIndex(TKey key) => (key.GetHashCode() & 0x7FFFFFFF) % _buckets.Length;
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private int GetEntryIndex(TKey key) => _buckets[GetBucketIndex(key)] - 1;
+
+        public int Count { get; private set; }
+
+        public bool ContainsKey(TKey key)
+        {
+            Entry[] entries = _entries;
+            int entryIndex = GetEntryIndex(key);
+
+            while (entryIndex != -1)
+            {
+                if (entries[entryIndex].key.Equals(key))
+                {
+                    return true;
+                }
+                entryIndex = entries[entryIndex].next;
+            }
+
+            return false;
+        }
+
+        public TValue GetValueOrDefault(TKey key)
+        {
+            Entry[] entries = _entries;
+            int entryIndex = GetEntryIndex(key);
+
+            while (entryIndex != -1)
+            {
+                if (entries[entryIndex].key.Equals(key))
+                {
+                    return entries[entryIndex].value;
+                }
+                entryIndex = entries[entryIndex].next;
+            }
+
+            return default;
+        }
+
+        public ref TValue this[TKey key]
+        {
+            get
+            {
+                Entry[] entries = _entries;
+                int entryIndex = GetEntryIndex(key);
+
+                while (entryIndex != -1)
+                {
+                    if (entries[entryIndex].key.Equals(key))
+                    {
+                        return ref entries[entryIndex].value;
+                    }
+                    entryIndex = entries[entryIndex].next;
+                }
+
+                if (Count == entries.Length) entries = Resize();
+
+                entryIndex = Count++;
+                entries[entryIndex].key = key;
+                int bucket = GetBucketIndex(key);
+                entries[entryIndex].next = _buckets[bucket] - 1;
+                _buckets[bucket] = entryIndex + 1;
+                return ref entries[entryIndex].value;
+            }
+        }
+
+        private Entry[] Resize()
+        {
+            var count = Count;
+            var newSize = HashHelpers.ExpandPrime(count);
+            var entries = new Entry[newSize];
+            Array.Copy(_entries, 0, entries, 0, count);
+            _entries = entries;
+
+            int[] newBuckets = new int[newSize];
+            _buckets = newBuckets;
+            for (int i = 0; i < count;)
+            {
+                int bucketIndex = GetBucketIndex(entries[i].key);
+                entries[i].next = newBuckets[bucketIndex] - 1;
+                newBuckets[bucketIndex] = ++i;
+            }
+
+            return entries;
+        }
+
+        public IEnumerable<TKey> Keys
+        {
+            get
+            {
+                var entries = _entries;
```

Comment:
nit: avoid using var here, and elsewhere. only use var where you can infer the type from the right hand side when reading the code.